src/components/ClassCardGrid.astro

```diff
@@ -1,18 +1,142 @@
+    due: localizeTime(entry.data.due),
+    notes: entry.data.notes,
+  },
+]);
+
+// Static sorting of upcoming events
+const upcomingLectures = lectures.filter(lecture => +new Date(lecture.date) > +new Date()).sort((a, b) => +new Date(a.date) - +new Date(b.date));
+const upcomingAssignments = assignments.filter(assignment => +new Date(assignment.due) > +new Date()).sort((a, b) => +new Date(a.due) - +new Date(b.due));
+const upcomingExams = exams.filter(exam => +new Date(exam.date) > +new Date()).sort((a, b) => +new Date(a.date) - +new Date(b.date));
 ---
 
 <CardGrid>
+  <!-- Lecture card -->
   <Card title="Upcoming Lecture" icon="pencil">
-    TODO: Show relevant information about the next lecture.
+    <div class="lecture-item" data-date={upcomingLectures[0]?.date}>
+      {upcomingLectures.length > 0 ? (
+        <>
+          The next lecture will cover: {upcomingLectures[0].title}. 
+          <div>Date: {formatTime(upcomingLectures[0].date)}</div>
+          <div>
+            <strong>Readings:</strong>
+            <ul>
+              {upcomingLectures[0].readings.map(reading => (
+                <li>
+                  <a href={reading.link} target="_blank" rel="noopener noreferrer">
+                    {reading.name}
+                  </a>
+                  {reading.optional && " (Optional)"}
+                </li>
+              ))}
+            </ul>
+          </div>
+        </>
+      ) : (
+        "There are no upcoming lectures."
+      )}
+    </div>
   </Card>
+
+  <!-- Assignment card -->
   <Card title="Upcoming Assignment" icon="laptop">
-    Show relevant information about the next assignment.
+    <div class="assignment-item" data-date={upcomingAssignments[0]?.due}>
+      {upcomingAssignments.length > 0 ? (
+        <>
+          <strong>Title:</strong> {upcomingAssignments[0].title}
+          <br />
+          <strong>Description:</strong> {upcomingAssignments[0].description}
+          <br />
+          <strong>Release Date:</strong> {formatTime(upcomingAssignments[0].release)}
+          <br />
+          <strong>Due Date:</strong> {formatTime(upcomingAssignments[0].due)}
+          <br />
+          Ensure that you submit via Gradescope before the deadline.
+        </>
+      ) : (
+        "No assignments are due soon."
+      )}
+    </div>
   </Card>
+
+  <!-- Exam card -->
   <Card title="Upcoming Exam" icon="open-book">
-    Show relevant information about the next exam.
+    <div class="exam-item" data-date={upcomingExams[0]?.date}>
+      {upcomingExams.length > 0 ? (
+        <>
+          The next exam is on: {formatTime(upcomingExams[0].date)}. Topics will include: {upcomingExams[0].content}.
+        </>
+      ) : (
+        "No upcoming exams scheduled."
+      )}
+    </div>
   </Card>
+
   <Card title="Explore Lecture Content" icon="document">
     Take a look at the lecture notes, slides, and course videos.
   </Card>
 </CardGrid>
+
+<!-- Client-side script to filter past events -->
```

Comment:
This still isn't correct unfortunately - the client side script should have access to the entire lecture list, and then dynamically render the correct upcoming content based on the current time. As is the .astro component will not update properly if there are two lectures' worth of content on the site and it isn't updated before vs. after a given lecture / other event happens.